### ISSUE TITLE
Add support for windows

### ISSUE THREAD

**Issue by kmaork** [kmaork/pyinjector#4]

**Comment by Scrooge20291224** [comment]
I tried to compile it with setup.py (with changed src dir "src/windows") but get a "stdio.h no found" error, and I tried nmake injector, and it succeded with several obj/lib/dll, how could I use these file to make accessible pyd for pyinjector? 

Thx!

**Comment by kmaork** [comment]
Hey @AlucardNosferatu , I've added windows support and uploaded windows wheels to pypi. Please check if they work for you :)

**Comment by Scrooge20291224** [comment]
> Hey @AlucardNosferatu , I've added windows support and uploaded windows wheels to pypi. Please check if they work for you :)

It's very nice of you that the update is so fast :) but as I try to use `pip install pyinjector`, I still cannot install it but with a new error:

```
C:\Windows\system32>pip install pyinjector
Looking in indexes: https://mirrors.aliyun.com/pypi/simple/
Collecting pyinjector
  Using cached https://mirrors.aliyun.com/pypi/packages/fa/86/95190dfe1b4fae06cd0bed1b7c90715f0d1950680ec346eaad3d3b31ad14/pyinjector-0.1.4.zip (23 kB)
Building wheels for collected packages: pyinjector
  Building wheel for pyinjector (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: 'c:\program files (x86)\microsoft visual studio\shared\python37_64\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"'; __file__='"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\16413\AppData\Local\Temp\pip-wheel-hnh0f1q4'
       cwd: C:\Users\16413\AppData\Local\Temp\pip-install-err9tpmn\pyinjector_2b66e69df7e840489404580dc2c6cff8\
  Complete output (20 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.7
  creating build\lib.win-amd64-3.7\pyinjector
  copying pyinjector\pyinjector.py -> build\lib.win-amd64-3.7\pyinjector
  copying pyinjector\__init__.py -> build\lib.win-amd64-3.7\pyinjector
  copying pyinjector\__main__.py -> build\lib.win-amd64-3.7\pyinjector
  running build_ext
  building 'pyinjector.libinjector' extension
  creating build\temp.win-amd64-3.7
  creating build\temp.win-amd64-3.7\Release
  creating build\temp.win-amd64-3.7\Release\injector
  creating build\temp.win-amd64-3.7\Release\injector\src
  creating build\temp.win-amd64-3.7\Release\injector\src\linux
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DEM_AARCH64=183 -Iinjector\include "-Ic:\program files (x86)\microsoft visual studio\shared\python37_64\include" "-Ic:\program files (x86)\microsoft visual studio\shared\python37_64\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /Tcinjector\src\linux\elf.c /Fobuild\temp.win-amd64-3.7\Release\injector\src\linux\elf.obj
  elf.c
  injector\src\linux\elf.c(27): fatal error C1083: 无法打开包括文件: “elf.h”: No such file or directory
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.27.29110\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
  ----------------------------------------
  ERROR: Failed building wheel for pyinjector
  Running setup.py clean for pyinjector
Failed to build pyinjector
Installing collected packages: pyinjector
    Running setup.py install for pyinjector ... error
    ERROR: Command errored out with exit status 1:
     command: 'c:\program files (x86)\microsoft visual studio\shared\python37_64\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"'; __file__='"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\16413\AppData\Local\Temp\pip-record-_8vt8p6p\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\program files (x86)\microsoft visual studio\shared\python37_64\Include\pyinjector'
         cwd: C:\Users\16413\AppData\Local\Temp\pip-install-err9tpmn\pyinjector_2b66e69df7e840489404580dc2c6cff8\
    Complete output (20 lines):
    running install
    running build
    running build_py
    creating build
    creating build\lib.win-amd64-3.7
    creating build\lib.win-amd64-3.7\pyinjector
    copying pyinjector\pyinjector.py -> build\lib.win-amd64-3.7\pyinjector
    copying pyinjector\__init__.py -> build\lib.win-amd64-3.7\pyinjector
    copying pyinjector\__main__.py -> build\lib.win-amd64-3.7\pyinjector
    running build_ext
    building 'pyinjector.libinjector' extension
    creating build\temp.win-amd64-3.7
    creating build\temp.win-amd64-3.7\Release
    creating build\temp.win-amd64-3.7\Release\injector
    creating build\temp.win-amd64-3.7\Release\injector\src
    creating build\temp.win-amd64-3.7\Release\injector\src\linux
    C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DEM_AARCH64=183 -Iinjector\include "-Ic:\program files (x86)\microsoft visual studio\shared\python37_64\include" "-Ic:\program files (x86)\microsoft visual studio\shared\python37_64\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /Tcinjector\src\linux\elf.c /Fobuild\temp.win-amd64-3.7\Release\injector\src\linux\elf.obj
    elf.c
    injector\src\linux\elf.c(27): fatal error C1083: 无法打开包括文件: “elf.h”: No such file or directory
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.27.29110\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
    ----------------------------------------
ERROR: Command errored out with exit status 1: 'c:\program files (x86)\microsoft visual studio\shared\python37_64\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"'; __file__='"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\16413\AppData\Local\Temp\pip-record-_8vt8p6p\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\program files (x86)\microsoft visual studio\shared\python37_64\Include\pyinjector' Check the logs for full command output.
```

from log I see the error is caused by miss-use of `elf.h` in linux src code, however, I checked your `setup.py` and see it already modified with switchable win32 src dir, maybe there is some kind of env miss-configure in my own PC, I will double-check it now.

**Comment by Scrooge20291224** [comment]
> > Hey @AlucardNosferatu , I've added windows support and uploaded windows wheels to pypi. Please check if they work for you :)
> 
> It's very nice of you that the update is so fast :) but as I try to use `pip install pyinjector`, I still cannot install it but with a new error:
> 
> ```
> C:\Windows\system32>pip install pyinjector
> Looking in indexes: https://mirrors.aliyun.com/pypi/simple/
> Collecting pyinjector
>   Using cached https://mirrors.aliyun.com/pypi/packages/fa/86/95190dfe1b4fae06cd0bed1b7c90715f0d1950680ec346eaad3d3b31ad14/pyinjector-0.1.4.zip (23 kB)
> Building wheels for collected packages: pyinjector
>   Building wheel for pyinjector (setup.py) ... error
>   ERROR: Command errored out with exit status 1:
>    command: 'c:\program files (x86)\microsoft visual studio\shared\python37_64\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"'; __file__='"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\16413\AppData\Local\Temp\pip-wheel-hnh0f1q4'
>        cwd: C:\Users\16413\AppData\Local\Temp\pip-install-err9tpmn\pyinjector_2b66e69df7e840489404580dc2c6cff8\
>   Complete output (20 lines):
>   running bdist_wheel
>   running build
>   running build_py
>   creating build
>   creating build\lib.win-amd64-3.7
>   creating build\lib.win-amd64-3.7\pyinjector
>   copying pyinjector\pyinjector.py -> build\lib.win-amd64-3.7\pyinjector
>   copying pyinjector\__init__.py -> build\lib.win-amd64-3.7\pyinjector
>   copying pyinjector\__main__.py -> build\lib.win-amd64-3.7\pyinjector
>   running build_ext
>   building 'pyinjector.libinjector' extension
>   creating build\temp.win-amd64-3.7
>   creating build\temp.win-amd64-3.7\Release
>   creating build\temp.win-amd64-3.7\Release\injector
>   creating build\temp.win-amd64-3.7\Release\injector\src
>   creating build\temp.win-amd64-3.7\Release\injector\src\linux
>   C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DEM_AARCH64=183 -Iinjector\include "-Ic:\program files (x86)\microsoft visual studio\shared\python37_64\include" "-Ic:\program files (x86)\microsoft visual studio\shared\python37_64\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /Tcinjector\src\linux\elf.c /Fobuild\temp.win-amd64-3.7\Release\injector\src\linux\elf.obj
>   elf.c
>   injector\src\linux\elf.c(27): fatal error C1083: 无法打开包括文件: “elf.h”: No such file or directory
>   error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.27.29110\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
>   ----------------------------------------
>   ERROR: Failed building wheel for pyinjector
>   Running setup.py clean for pyinjector
> Failed to build pyinjector
> Installing collected packages: pyinjector
>     Running setup.py install for pyinjector ... error
>     ERROR: Command errored out with exit status 1:
>      command: 'c:\program files (x86)\microsoft visual studio\shared\python37_64\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"'; __file__='"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\16413\AppData\Local\Temp\pip-record-_8vt8p6p\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\program files (x86)\microsoft visual studio\shared\python37_64\Include\pyinjector'
>          cwd: C:\Users\16413\AppData\Local\Temp\pip-install-err9tpmn\pyinjector_2b66e69df7e840489404580dc2c6cff8\
>     Complete output (20 lines):
>     running install
>     running build
>     running build_py
>     creating build
>     creating build\lib.win-amd64-3.7
>     creating build\lib.win-amd64-3.7\pyinjector
>     copying pyinjector\pyinjector.py -> build\lib.win-amd64-3.7\pyinjector
>     copying pyinjector\__init__.py -> build\lib.win-amd64-3.7\pyinjector
>     copying pyinjector\__main__.py -> build\lib.win-amd64-3.7\pyinjector
>     running build_ext
>     building 'pyinjector.libinjector' extension
>     creating build\temp.win-amd64-3.7
>     creating build\temp.win-amd64-3.7\Release
>     creating build\temp.win-amd64-3.7\Release\injector
>     creating build\temp.win-amd64-3.7\Release\injector\src
>     creating build\temp.win-amd64-3.7\Release\injector\src\linux
>     C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DEM_AARCH64=183 -Iinjector\include "-Ic:\program files (x86)\microsoft visual studio\shared\python37_64\include" "-Ic:\program files (x86)\microsoft visual studio\shared\python37_64\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\include" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\winrt" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\cppwinrt" /Tcinjector\src\linux\elf.c /Fobuild\temp.win-amd64-3.7\Release\injector\src\linux\elf.obj
>     elf.c
>     injector\src\linux\elf.c(27): fatal error C1083: 无法打开包括文件: “elf.h”: No such file or directory
>     error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.27.29110\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
>     ----------------------------------------
> ERROR: Command errored out with exit status 1: 'c:\program files (x86)\microsoft visual studio\shared\python37_64\python.exe' -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"'; __file__='"'"'C:\\Users\\16413\\AppData\\Local\\Temp\\pip-install-err9tpmn\\pyinjector_2b66e69df7e840489404580dc2c6cff8\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record 'C:\Users\16413\AppData\Local\Temp\pip-record-_8vt8p6p\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\program files (x86)\microsoft visual studio\shared\python37_64\Include\pyinjector' Check the logs for full command output.
> ```
> 
> from log I see the error is caused by miss-use of `elf.h` in linux src code, however, I checked your `setup.py` and see it already modified with switchable win32 src dir, maybe there is some kind of env miss-configure in my own PC, I will double-check it now.

I can successfully install it with downloaded zip and `python setup.py install`, but not `pip install`.

I successfully inject the c code given in tests, though there is some flaws, but the 'hello world' still successfully showed up in the target process, I guess the only problem is, how I could inject some py into another py? or make hypno accessible on win32 platform...?

Well that's enough for me anyway. Thank you ! @kmaork 👍 👍

**Comment by kmaork** [comment]
The reason it failed for you is because you are using a mirror of Pypi (mirrors.aliyun.com) and not pypi itself. You can see in the logs that it used `pyinjector==0.1.4`, and not `1.0.1`. That is an older version that does not support windows.
I don't know this mirror, but it will probably sync with Pypi soon and you will be able to install the `whl` files I uploaded, instead of building it youself.

And you're welcome, hope you enjoy using the library :)
Regarding injection of python code, I have opened https://github.com/kmaork/hypno/issues/3. I might get to do it this week but can't promise :)

**Comment by kmaork** [comment]
I now released `hypno==1.0.0` which supports Windows and uploaded windows wheels to pypi 😄
